Closes #347: ADRS with Lugano 2014 Criteria [create website] (#353)
pharmaverse/admiralonco: vignettes/adrs_lymphoma.Rmd

Note that only the Lugano 2014-specific derivation of the combined overall timepoint response is covered in this vignette, specifically for scenarios where the combined overall response is not collected directly on the CRF. For extended guidance on common steps in ADRS creation and additional response endpoints, refer to the examples in [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

# Lugano 2014 Response Categories for Lymphoma

The Lugano 2014 response criteria define lymphoma response using PET-CT based metabolic assessment and CT-based anatomic assessment.

The following tables summarize the response categories used in this vignette for PET-CT and CT assessments. These summaries are intended to support the example derivations and should be aligned with the study protocol and statistical analysis plan.

## PET-CT Based Response Categories

```{r pet-response-categories, echo=FALSE, message=FALSE}
pet_response_categories <- tibble::tribble(
  ~`PET-CT Response`, ~`Description`, ~`Derived Overall Category`,
  "CMR", "Complete metabolic response", "CR",
  "PMR", "Partial metabolic response", "PR",

>   "SMD", "Stable metabolic disease", "SD",

Reviewer: @kiran-peddamudium SMD should be "Stable metabolic disease" or "Stable Anatomic response" . Roche specs Stable Anatomic response" see below.	
Abbreviation	Meaning	
CMR / CAR / CR	Complete Metabolic Response / Complete Anatomic Response / Complete Response	
PMR / PAR / PR	Partial Metabolic Response / Partial Anatomic Response / Partial Response	
NMR / SAD / SD	No Metabolic Response / Stable Anatomic Response / Stable Disease 	
PMD / PAD / PD	Progressive Metabolic Disease / Progressive Anatomic Disease /Progressive Disease	
PSP	Pseudoprogression	
NED	No Evidence of Disease (used by BICR only)

  "NMR", "No metabolic response", "SD",
  "PMD", "Progressive metabolic disease", "PD",
  "NE",  "Not evaluable", "NE or fallback logic",
  "ND",  "Not done or not determined", "Fallback to CT if available"
)

pet_response_categories |>
  gt::gt() |>
  gt::tab_header(
    title = "Table 1: PET-CT Based Response Categories",
    subtitle = "Lugano 2014 response categories used in this vignette"
  )
```

## CT-Based Response Categories

```{r ct-response-categories, echo=FALSE, message=FALSE}

> ct_response_categories <- tibble::tribble(

Reviewer: @kiran-peddamudium CT - anatomical response are given in Roche specs as CAR,PAR,SAD, PAD and PTA+CT is CR, PR.. you are giving CT is CR, PR..?WHY?

Reviewer: as discussed in the meeting , seems like updated

  ~`CT Response`, ~`Description`, ~`Derived Overall Category`,
  "CR", "Complete response", "CR",
  "PR", "Partial response", "PR",
  "SD", "Stable disease", "SD",
  "PD", "Progressive disease", "PD",
  "NE", "Not evaluable", "NE",
  "ND", "Not done or not determined", "ND",
  "NED", "No evidence of disease", "NED"
)

ct_response_categories |>
  gt::gt() |>
  gt::tab_header(
    title = "Table 2: CT-Based Response Categories",
    subtitle = "Anatomic response categories used in this vignette"
  )
```

In the example data, `SMD` is used for stable metabolic disease. Some implementations may use `NMR` for no metabolic response. For the purpose of the combined overall response derivation in this vignette, both `SMD` and `NMR` are mapped to `SD`.

> Values such as `NED`, `PSP`, and `ND` may require study-specific handling. The simplified derivation implemented in this vignette focuses on deriving a combined overall response from available PET-CT and CT response records.

Reviewer: Can you define PSP also in the PET response table above.

Reviewer 2: Remove PSP from code.

Reviewer 2: Done

  collapse = TRUE,
  comment = "#>"
)
library(admiraldev)
library(gt)
# TO BE REMOVED ONCE THE BRANCH IS MERGED TO MAIN
remotes::install_github("pharmaverse/pharmaversesdtm", ref = "235-general-issue-add-rs-data-for-lymphoma")
```

# Introduction

This article describes creating an `ADRS` ADaM dataset for lymphoma studies based on the [**Lugano 2014 response criteria**](https://doi.org/10.1200/JCO.2013.54.8800).

Lymphoma response assessment under Lugano 2014 is based on a combination of imaging-based evaluations:

-   **PET-CT based assessment**, providing metabolic response evaluation using the 5-point Deauville scale.
-   **CT-based assessment**, providing anatomic response evaluation of nodal and extranodal disease.

Depending on the study and the disease subtype, response evaluation may use PET-CT as the primary modality (for FDG-avid lymphomas) or CT alone (for non–FDG-avid lymphomas). Some studies collect both PET-CT and CT response components, and a combined overall response is derived.

Please check [Lugano 2014 Classification](https://imagingendpoints.com/wp-content/uploads/2022/07/IEP-6936-Lugano-at-IE-2022-FV1.0_DIGITAL_Final-25-07-22.pdf) for more details.

> Note that only the Lugano 2014-specific derivation of the combined overall timepoint response is covered in this vignette, specifically for scenarios where the combined overall response is not collected directly on the CRF. For extended guidance on common steps in ADRS creation and additional response endpoints, refer to the examples in [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

Reviewer 3: I think this is a very important call out. From some brief across industry discussions, it seems that some companies will put the onus on the investigator to provide an overall response by Lugano which makes perfect sense to me. In this case, everything below becomes more or less redundant so perhaps we can emphasize this point more (add in bold perhaps?). 

I'd even go further to clarify that everything below is just an example of how the integrated time-point response could be derived and is not intended to be instructions/guidance. I say this because I have worked on numerous Lugano 2014 based studies and there have always been slight variations between the studies in how this was handled. I think this would be worthwhile to point out.

# Lugano 2014 Response Categories for Lymphoma

The Lugano 2014 response criteria define lymphoma response using PET-CT based metabolic assessment and CT-based anatomic assessment.

The following tables summarize the response categories used in this vignette for PET-CT and CT assessments. These summaries are intended to support the example derivations and should be aligned with the study protocol and statistical analysis plan.

## PET-CT Based Response Categories

```{r pet-response-categories, echo=FALSE, message=FALSE}
pet_response_categories <- tibble::tribble(
  ~`PET-CT Response`, ~`Description`, ~`Derived Overall Category`,
  "CMR", "Complete metabolic response", "CR",
  "PMR", "Partial metabolic response", "PR",
  "SMD", "Stable metabolic disease", "SD",

>   "NMR", "No metabolic response", "SD",

Reviewer 3: I'd consider putting SMD and NMR on the same row simply because these two terms are interchangeable. My understanding is that, in the context of a PET-CT scan, NMR is the more technically correct term but sometimes SMD is used to align closer with the anatomic response by CT.

  "PMD", "Progressive metabolic disease", "PD",
  "NE",  "Not evaluable", "NE or fallback logic",

>   "ND",  "Not done or not determined", "Fallback to CT if available"

Reviewer 3: Consider adding the case of NED by PET-CT. A response of NED would only ever be reported by an IRC/IRF, not the investigator and refers to a situation where the IRC receives a baseline PET scan, but does not identify any FDG-avid disease on that scan. Of course, if this is the case, all PET-CT scans for this patient will have a response of NED as they have no baseline scan to compare back to. In my prior experience, if a response of NED is reported, the integrated-time-point response defaults to the CT scan. The IRC/IRF may still be able to provide a response by CT scan because it might just be that they can identify a target/non-target lesion but it is not FDG-avid which means it wouldn't appear on the PET-CT scan. If both the PET-CT scan and CT scan is NED, then the overall ITPR is usually set to NED.

-   [Derive Combined Overall Timepoint Response (OVRLRESC)](#ovrlresc)
-   [Other Endpoints](#other)

## Required Packages

The examples of this vignette require the following packages.

```{r, warning=FALSE, message=FALSE}
library(admiral)
library(admiralonco)
library(pharmaversesdtm)
library(pharmaverseadam)
library(dplyr)
library(tibble)
```

## Read in Data {#readdata}

To begin, all data frames needed for the creation of `ADRS` should be read into the environment. This will be a company-specific process. For this vignette, the main input datasets are `ADSL` and `RS`.

For demonstration purposes, the SDTM and ADaM datasets based on CDISC Pilot test data from `{pharmaversesdtm}` and `{pharmaverseadam}` are used.

In this vignette, the `RS` SDTM dataset is expected to contain lymphoma response assessments based on Lugano 2014 criteria. The example `RS` dataset contains separate records for:

-   PET-CT based response assessments, identified by `RSSCAT = "INCLUDING PET-CT SCAN"`.
-   CT-based response assessments, identified by `RSSCAT = "NOT INCLUDING PET SCAN"`.

```{r message=FALSE}

# Lymphoma SDTM data
#rs <- pharmaversesdtm::rs_onco_lymphoma

> rs <- rs_onco_lymphoma

Reviewer 3: Even when I load the 235-general-issue-add-rs-data-for-lymphoma branch of the pharamversesdtm repo, rs_onco_lymphoma does not appear to be an exported object so this will need updating. All testing below is performed by manually re-creating the pharamversesdtm data by running the code within the 235-general-issue-add-rs-data-for-lymphoma branch of the pharamversesdtm repo.

Reviewer 2: @reesnj , when did you try it?

I've updated the feature branch in pharmaversesdtm yesterday because there was a bug which prevented loading the package.

I've also updated this vignette such that it installs pharmaversesdtm from the feature branch.

Reviewer 3: This is working now - thanks @bundfussr!

# Convert blanks to NA
rs <- convert_blanks_to_na(rs)

# ADaM data
adsl <- pharmaverseadam::adsl
```

```{r echo=FALSE}
# select subjects from adsl such that there is one subject without RS data
rs_subjects <- unique(rs$USUBJID)
adsl_subjects <- unique(adsl$USUBJID)
adsl <- filter(
  adsl,
  USUBJID %in% union(rs_subjects, setdiff(adsl_subjects, rs_subjects)[1])
)
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
   rs,
  display_vars = exprs(USUBJID, RSTESTCD,RSCAT,RSSCAT,RSMETHOD, RSSTRESC, VISIT, VISITNUM, RSDTC)
)
```

## Pre-processing of Input Records {#input}

At this step, it may be useful to join `ADSL` to your `RS` domain. Only the `ADSL` variables used for derivations are selected at this step.

```{r eval=TRUE}
adsl_vars <- exprs(TRTSDT)
adrs <- derive_vars_merged(
  rs,
  dataset_add = adsl,
  new_vars = adsl_vars,
  by_vars = get_admiral_option("subject_keys")
)

```

### Partial Date Imputation and Deriving `ADT`, `ADTF`, `AVISIT`, `AVISITN` etc.

If your data collection allows for partial dates, you could apply a company-specific imputation rule at this stage when deriving `ADT`. For this example, here we impute missing day to last possible date.

```{r}

> adrs <- adrs %>%

Reviewer 3: If this is standard practice in `admiralonco` please ignore me but I'd personally prefer not to create a new dataset with the same name here as it makes debugging much harder.

Reviewer 2: Yes, in the admiral open source packages we use the same name. It is unlikely that the code provided by vignettes and templates in the open source packages is used without major updates. Modifying the code is easier when the same name is used (and it is more readable in my opion).

Reviewer 3: Sure, no problem. This is usually something I personally try to avoid but happy to keep this as is if it's consistent with admiral standard practice.

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Prepare param_lookup for SDTM RSTESTCD and RSSCAT to add metadata
param_lookup <- tibble::tribble(
  ~RSTESTCD,  ~RSSCAT,                   ~PARAMCD, ~PARAM,          ~PARAMN,
  "OVRLRESP", "INCLUDING PET-CT SCAN",   "PETRSP", "PET-CT Response", 1,
  "OVRLRESP", "NOT INCLUDING PET SCAN",  "CTRSP",  "CT Response",     2
)

adrs <- adrs %>%
  derive_vars_merged_lookup(
    dataset_add = param_lookup,
    by_vars = exprs(RSTESTCD, RSSCAT)
  ) %>%
  mutate(
    PARCAT1 = RSCAT,
    AVALC = case_when(
      RSSTAT == "NOT DONE" ~ "ND",
      TRUE ~ RSSTRESC
    )
  )
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
   adrs,
  display_vars = exprs(USUBJID,PARAMCD,PARAM,AVALC, AVISIT, AVISITN,ADT)
)
```

## Derive Combined Overall Timepoint Response (`OVRLRESC`) Parameter {#ovrlresc}

For this vignette, the combined overall timepoint response parameter, `OVRLRESC`, is derived from the PET-CT and CT response records collected at each visit.

> This example represents a scenario where the combined overall response is `not collected directly` on the CRF. Instead, it is derived using the available PET-CT and CT response records.

Reviewer 3: Did you mean to put "not collected directly" in bold here rather than in backticks?

### General Derivation Assumptions Used in This Vignette

The following table summarizes the assumptions used in this vignette to derive the combined overall timepoint response from PET-CT and CT response records under Lugano 2014. These assumptions are intended for demonstration purposes. Please refer to your study protocol, statistical analysis plan, and other study documentation before using in production analyses.

#### Table: Combined Overall Timepoint Response Based on Lugano 2014 Response Categories

```{r , eval=TRUE, include=TRUE, message=FALSE, echo=FALSE}
combined_lugano_table <- tibble::tribble(
  ~`PET-CT Response`,                         ~`CT Response`,            ~`Combined Overall Response`,
  "CMR",                                      "Any",                     "CR",
  "PMR",                                      "Any",                     "PR",

>   "SMD or NMR",                               "Any non-progressive",     "SD",

Reviewer 3: I think the CT Response here should read "Any". In the situation where we have a "NMR" by PET-CT, the overall response would always be "SD" regardless of the CT response.

  "PMD",                                      "Any",                     "PD",
  "NE",                                       "NE",                      "NE",

>   "NE, with prior evaluable PET-CT",          "Non-progressive",         "Carry forward prior PET-CT response",

Reviewer 3: In the CT response here, we say "Non-progressive" but in the row below, we actually call out all the possible values which appear to all be non-progressive, so could we be consistent here to avoid any confusion? Either always call out exactly which values or just group as "Non-progressive". I'd probably prefer we call out the values with the exception of the "Any" case.

  "NE, no prior evaluable PET-CT",            "CR / PR / SD / NED",      "Use current CT response",
  "NE",                                       "PD",                      "PD",

>   "NED",                                      "CR / PR / SD / PD / NE",  "NED",

Reviewer 3: In the situation where the IRC report a response of "NED" because the baseline scan shows no evidence of FDG-Avid disease, but there do report a value by CT scan, we would usually default to the CT outcome.

  "NE",                                       "NED",                     "NED",
  "NED",                                      "NED",                     "NED",

>   "ND or not done",                           "CR / PR / SD / PD / NED", "Use current CT response",

Reviewer 3: In my prior experience, I have seen NEs and NDs handled consistently. So, there wouldn't be a special rule for ND that is different from NE. Perhaps we could update all rows above to say "NE / ND" instead? For example, say we had a PET-CT which wasn't done (ND) and a CT scan that was PR, if the previous PET-CT scan was CMR, we would carry this forward (i.e; set overall response to CR). That wouldn't be clear to me from looking at this table.

  ) |>
  gt::tab_source_note(
    source_note = "Pseudoprogression handling, including final-visit conversion, is study-specific and not implemented in this example."
  )
```

##### Combined Overall Time Point Response(`OVRLRESC`) Records referenced from above table.

###### Derive prior evaluable PET-CT response for carry-forward logic

```{r, eval=TRUE, message=FALSE, include=TRUE}

adrs_prior_pet <- adrs %>%
  filter(PARAMCD == "PETRSP") %>%
  derive_vars_joined(
    dataset_add = adrs %>%
      filter(PARAMCD == "PETRSP",
             AVALC %in% c("CMR", "PMR", "SMD", "NMR", "PMD", "NED", "PSP")) %>%
      select(!!!get_admiral_option("subject_keys"),ADT,AVISITN,AVALC) %>%
      distinct(),
    by_vars = get_admiral_option("subject_keys"),
    order = exprs(ADT, AVISITN),
    mode = "last",
    join_type = "before",
    filter_join = ADT.join < ADT,
    new_vars = exprs(
     AVALC_P = AVALC,
     ADT_P = ADT
    )
  )

adrs <- bind_rows(
  adrs_prior_pet,
  adrs %>% filter(PARAMCD != "PETRSP")
)
```

###### Derive Combined Overall Time Point Response.

```{r eval=TRUE, message=FALSE, include=TRUE}
adrs <- derive_param_computed(
  dataset = adrs,

>   by_vars = exprs(!!!get_admiral_option("subject_keys"), !!!adsl_vars, DOMAIN, ADT, ADY, ADTM, ADTF, VISIT, VISITNUM, AVISIT, AVISITN),

Reviewer 3: Usually, I've seen RSSPID used to link the PET-CT and CT records which usually refers to these records being entered together on the CRF. Therefore, even if the PET-CT scan and CT scan were done a few days apart, we'd still derive an overall time-point response. In terms of the date that is chosen for the time-point response, my previous experience is usually that the date from whichever record is chosen is selected, unless both the PET-CT and CT scan show progressive disease, in which case the earliest is picked.

Author: Thank you Nathan, will be discussed in upcoming meeting

Reviewer 2: `RSSPID` is not in the test data. Thus we should just add a note to explain the issue and that the `by_vars` may need to be needed to be updated depending on the data collection.

Reviewer 2: @reesnj , I've added a note regarding `RSSPID`. Please have a look.

    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "NE", "ND", "NED") ~
    map_pet_to_overall(AVALC_P.PETRSP),

  # PET-CT is NE or ND and no prior evaluable PET-CT exists
  AVALC.PETRSP %in% c("NE", "ND") &
    is.na(AVALC_P.PETRSP) &
    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "PAD", "NED") ~
    map_ct_to_overall(AVALC.CTRSP),

  # PET-CT is NE and CT is also NE or ND or missing
  AVALC.PETRSP == "NE" &
    (AVALC.CTRSP %in% c("NE", "ND") | is.na(AVALC.CTRSP)) ~
    "NE",

  # PET-CT is ND and CT is also NE or ND or missing
  AVALC.PETRSP == "ND" &
    (AVALC.CTRSP %in% c("NE", "ND") | is.na(AVALC.CTRSP)) ~
    "ND",

  # PET-CT missing; use CT response if available
  is.na(AVALC.PETRSP) &
    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "PAD", "NED", "NE", "ND") ~
    map_ct_to_overall(AVALC.CTRSP),

  # No valid response available
  TRUE ~ "ND"),
    PARAMCD = "OVRLRESC",
    PARAM = "Overall Response - Derived",
    PARAMN = 3,
    PARCAT1 = "LUGANO 2014"
  )
)
```

```{r, echo=FALSE}
plot_response <- function(dataset, page, page_size = 4) {
  subjects <- unique(dataset$USUBJID)
  page_nrs <- subjects %>% rank() %>% `/`(page_size) %>% ceiling()
  
  ggplot(dataset %>%
         mutate(

>            page_nr = recode_values(USUBJID, from = subjects, to = page_nrs),

Author: @bundfussr Hi Stefan,
The visualization in the vignette is very useful and well presented. However, I am encountering an issue when running the code like couldn't find function recode_values(), even though dplyr is loaded in the session.
Could you please advise if this function belongs to a specific package version or if any additional package needs to be loaded upstream?

Reviewer 2: Hi @kiran-peddamudium , `recode_values()` was added to `dplyr` in the latest version (1.2.0).

The derivation below is provided only as an example of how an integrated timepoint response could be derived when the combined overall response is **not collected directly**. It is not intended as general Lugano 2014 implementation guidance. Study-specific rules may vary and should be aligned with the protocol, SAP, CRF design, and data review conventions.

For extended guidance on common steps in ADRS creation and additional response endpoints, refer to the examples in [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

# Lugano 2014 Response Categories for Lymphoma

The Lugano 2014 response criteria define lymphoma response using PET-CT based metabolic assessment and CT-based anatomic assessment.

The following tables summarize the response categories used in this vignette for PET-CT and CT assessments. These summaries are intended to support the example derivations and should be aligned with the study protocol and statistical analysis plan.

## PET-CT Based Response Categories

```{r pet-response-categories, echo=FALSE, message=FALSE}
pet_response_categories <- tibble::tribble(

>   ~`PET-CT Response`, ~`Description`, ~`Derived Overall Category`,

Reviewer 2: What is the "Derived Overall Category"?

Does it describe the combination of PET and CT or an intermediate result?

Reviewer 2: Remove column in both tables.

#### Table: Combined Overall Timepoint Response Based on Lugano 2014 Response Categories

```{r , eval=TRUE, include=TRUE, message=FALSE, echo=FALSE}
combined_lugano_table <- tibble::tribble(
  ~`PET-CT Response`,                         ~`CT Response`,                     ~`Combined Overall Response`,
  "CMR",                                      "Any",                              "CR",
  "PMR",                                      "Any",                              "PR",
  "NMR or SMD",                               "Any",                              "SD",
  "PMD",                                      "Any",                              "PD",
  "PSP",                                      "Any",                              "PSP",
  "NED",                                      "CAR / PAR / SAD / PAD / NE / ND",  "Use current CT response",

>   "NED",                                      "NED",                              "NED",

Reviewer 2: Could these two lines simplified to

```suggestion
  "NED",                                      Any,                                "Use current CT response",
```

  "NE / ND, with prior evaluable PET-CT",     "CAR / PAR / SAD / NE / ND / NED",  "Carry forward prior PET-CT response",
  "NE / ND, with prior evaluable PET-CT",     "PAD",                              "PD",
  "NE / ND, no prior evaluable PET-CT",       "CAR / PAR / SAD / PAD / NED",      "Use current CT response",
  "NE / ND, no prior evaluable PET-CT",       "NE / ND / Missing",                "NE or ND",
  "Missing",                                  "CAR / PAR / SAD / PAD / NED",      "Use current CT response",

>   "Missing",                                  "NE / ND",                          "Use current CT response",

Reviewer 2: Could this be simplified to
```suggestion
  "Missing",                                  Any,                                "Use current CT response",
```
?

    subtitle = "PET-CT, CT, and Combined Overall Response Mapping"
  ) |>
  gt::tab_source_note(
    source_note = "This table is example-only and should be aligned with the study protocol and statistical analysis plan."
  ) |>
  gt::tab_source_note(
    source_note = "For evaluable PET-CT responses CMR, PMR, NMR or SMD, and PMD, the PET-CT response determines the integrated response in this example."
  ) |>
  gt::tab_source_note(
    source_note = "When PET-CT is NE or ND and the current CT response is not progressive, prior evaluable PET-CT response may be carried forward."
  ) |>
  gt::tab_source_note(
    source_note = "When PET-CT is reported as NED, the integrated response generally defaults to the CT response if available. If both PET-CT and CT are NED, the integrated response is NED."
  ) |>
  gt::tab_source_note(
    source_note = "Pseudoprogression handling is study-specific and is not implemented further in this example."
  )
```

##### Combined Overall Timepoint Response(`OVRLRESC`) Records referenced from above table.

```{r, eval=TRUE, include=TRUE, message=FALSE}
# Pre-processing for Overall values
map_pet_to_overall <- function(x) {

>   case_when(

Reviewer 2: This could be simplified by using `recode_values()`.

    x == "CMR" ~ "CR",
    x == "PMR" ~ "PR",
    x %in% c("NMR", "SMD") ~ "SD",
    x == "PMD" ~ "PD",
    x == "NED" ~ "NED",
    x == "PSP" ~ "PSP",
    TRUE ~ NA_character_
  )
}

map_ct_to_overall <- function(x) {

>   case_when(

Reviewer 2: This could be simplified by using `recode_values()`.

    x == "CAR" ~ "CR",
    x == "PAR" ~ "PR",
    x == "SAD" ~ "SD",
    x == "PAD" ~ "PD",
    x == "NED" ~ "NED",
    x == "NE" ~ "NE",
    x == "ND" ~ "ND",
    TRUE ~ NA_character_
  )
}
```

###### Derive prior evaluable PET-CT response for carry-forward logic

```{r, eval=TRUE, message=FALSE, include=TRUE}

adrs_prior_pet <- adrs %>%
  filter(PARAMCD == "PETRSP") %>%
  derive_vars_joined(
    dataset_add = adrs %>%
      filter(
        PARAMCD == "PETRSP",

>         AVALC %in% c("CMR", "PMR", "NMR", "SMD", "PMD", "NED", "PSP")

Reviewer 2: I think we shouldn't carry forward `"NED"`.

I'm not sure regarding `"PSP"`.

Reviewer 2: Remove `"NED"` keep `"PSP"`

Reviewer 2: Removed both according to decision below.

      ) %>%
      select(!!!get_admiral_option("subject_keys"), ADT, AVISITN, AVALC) %>%

>       distinct(),

Reviewer 2: Do we need this? I wouldn't expect duplicates.

Reviewer 2: Remove it

    by_vars = get_admiral_option("subject_keys"),
    order = exprs(ADT, AVISITN),
    mode = "last",
    join_type = "before",
    filter_join = ADT.join < ADT,
    new_vars = exprs(
      AVALC_P = AVALC,
      ADT_P = ADT
    )
  )

adrs <- bind_rows(
  adrs_prior_pet,
  adrs %>% filter(PARAMCD != "PETRSP")

> )

Reviewer 2: ```suggestion
  restrict_derivation(
    filter = PARAMCD == "PETRSP",
    derivation = derive_vars_joined,
    args = params(
      dataset_add = adrs %>%
        filter(
          PARAMCD == "PETRSP",
          AVALC %in% c("CMR", "PMR", "NMR", "SMD", "PMD", "NED", "PSP")
        ) %>%
        select(!!!get_admiral_option("subject_keys"), ADT, AVISITN, AVALC) %>%
        distinct(),
      by_vars = get_admiral_option("subject_keys"),
      order = exprs(ADT, AVISITN),
      mode = "last",
      join_type = "before",
      filter_join = ADT.join < ADT,
      new_vars = exprs(
        AVALC_P = AVALC,
        ADT_P = ADT
      )
    )
  )
```

  # PET-CT NED and CT missing: keep NED
  AVALC.PETRSP == "NED" & is.na(AVALC.CTRSP) ~
    "NED",

  # PET-CT is NE or ND and CT indicates progression
  AVALC.PETRSP %in% c("NE", "ND") & AVALC.CTRSP == "PAD" ~
    "PD",

  # PET-CT is NE or ND and prior evaluable PET-CT exists
  AVALC.PETRSP %in% c("NE", "ND") &
    !is.na(AVALC_P.PETRSP) &
    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "NE", "ND", "NED") ~
    map_pet_to_overall(AVALC_P.PETRSP),

  # PET-CT is NE or ND and no prior evaluable PET-CT exists
  AVALC.PETRSP %in% c("NE", "ND") &
    is.na(AVALC_P.PETRSP) &
    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "PAD", "NED") ~
    map_ct_to_overall(AVALC.CTRSP),

  # PET-CT is NE and CT is also NE or ND or missing
  AVALC.PETRSP == "NE" &
    (AVALC.CTRSP %in% c("NE", "ND") | is.na(AVALC.CTRSP)) ~
    "NE",

  # PET-CT is ND and CT is also NE or ND or missing
  AVALC.PETRSP == "ND" &
    (AVALC.CTRSP %in% c("NE", "ND") | is.na(AVALC.CTRSP)) ~
    "ND",

  # PET-CT missing; use CT response if available
  is.na(AVALC.PETRSP) &
    AVALC.CTRSP %in% c("CAR", "PAR", "SAD", "PAD", "NED", "NE", "ND") ~
    map_ct_to_overall(AVALC.CTRSP),

  # No valid response available

>   TRUE ~ "ND"),

Reviewer 2: Should this cover the case that neither PET nor CT scan is available?

If we want to cover this case, we need to add records first.

Reviewer 2: No imputation needed

           ),
           Visit = factor(
             AVISIT,
             levels = c("BASELINE", "WEEK 8", "WEEK 16", "WEEK 24")
           )
         ) %>% 
         filter(page_nr == !!page)) +
  theme_classic() +
  theme(
    strip.background = element_blank(),
    axis.line.y = element_blank(),
    axis.ticks.y = element_blank(),
    legend.background = element_rect(fill = "whitesmoke")
  ) +
  geom_point(
    aes(
      x = Visit,
      y = Parameter,
      shape = Response,
      color = Response,
      fill = Response
    ),
    size = 4
  ) +
  scale_color_manual(
    values = c("CMR, CAR, CR" = "darkgreen", "PMR, PAR, PR" = "orange", "NMR, SAD, SD" = "blue", "PMD, PAD, PD" = "red", "NE" = "gray", "ND" = "black"), na.value = "lightgray"
  ) +
  scale_fill_manual(
    values = c("CMR, CAR, CR" = "darkgreen", "PMR, PAR, PR" = "orange", "NMR, SAD, SD" = "blue", "PMD, PAD, PD" = "red", "NE" = "gray", "ND" = "black"), na.value = "lightgray"
  ) +
  scale_shape_manual(
    values = c("CMR, CAR, CR" = 17, "PMR, PAR, PR" = 18, "NMR, SAD, SD" = 16, "PMD, PAD, PD" = 25, "NE" = 3, "ND" = 4),
    na.value = 22
  ) +
  facet_wrap(vars(paste("Subject", USUBJID)), ncol = 1)
}

> for (i in 1:ceiling(length(unique(adrs$USUBJID)) / 4)) {

Reviewer 2: Maybe we shouldn't display the visualization for all subjects but just show some of the more interesting scenarios.

What do you think?

Reviewer 2: Keep them all

  print(plot_response(adrs, page = i))
}
```

```{r, eval=TRUE, echo=FALSE}
dataset_vignette(
  adrs %>%
    filter(PARAMCD %in% c("OVRLRESC")) %>%
    arrange(!!!get_admiral_option("subject_keys"), AVISITN, PARAMN),
  display_vars = exprs(USUBJID, PARAMCD, PARAM, PARCAT1, AVALC, AVISIT, ADT)
)
```

### Derive `AVAL` (Numeric tumor response from `AVALC` values)

The `AVAL` values are not considered in the further parameter derivations below, and so changing `AVAL` here would not change the result of those derivations.

```{r}
adrs <- adrs %>%
  mutate(

>     AVAL = case_when(

Reviewer 2: `recode_values()` could be used here as well.

      AVALC %in% c("CR", "CMR", "CAR") ~ 1,
      AVALC %in% c("PR", "PMR", "PAR") ~ 2,
      AVALC %in% c("SD", "NMR", "SMD", "SAD") ~ 3,
      AVALC %in% c("PD", "PMD", "PAD") ~ 4,
      AVALC == "NE" ~ 5,
      AVALC == "NED" ~ 6,
      AVALC == "PSP" ~ 7,
      AVALC == "ND" ~ 8,
      TRUE ~ NA_real_
    )
  )
```

```{r, echo=FALSE}
dataset_vignette(
  adrs %>%
    filter(PARAMCD %in% c("OVRLRESC")) %>%
    arrange(!!!get_admiral_option("subject_keys"), AVISITN, PARAMN),
  display_vars = exprs(USUBJID, PARAMCD, PARAM, PARCAT1, AVALC, AVAL, AVISIT, ADT)
)
```

## Other Endpoints {#other}

> For examples of other endpoints, such as Best Overall Response (BOR), Confirmed Best Overall Response (CBOR), and other oncology response endpoints, please see [Creating ADRS (Including Non-standard Endpoints)](adrs.html).

Reviewer 2: As `OVRLRESC` is just an intermediate endpoint which is used to derive the endpoints which are summarized in the TLGs, I wonder if we should rephrase this paragraph to something like

```suggestion
The `OVRLRESC` parameter can be used as input for the derivation of standard endpoints, such as Best Overall Response (BOR), Confirmed Best Overall Response (CBOR), and other oncology response endpoints. Please see [Creating ADRS (Including Non-standard Endpoints)](adrs.html) for guidance on how to derive them.
```

Reviewer: okay with the suggestions..
